File: binding/src/libqstringconverter_base.cpp
#include <QStringConverter>
#include <QStringConverterBase>
#define WORKAROUND_INNER_CLASS_DEFINITION_QStringConverterBase__State
#include <qstringconverter_base.h>
#include "libqstringconverter_base.h"
#include "libqstringconverter_base.hxx"

bool QStringConverter_IsValid(const QStringConverter* self) {
	return self->isValid();
}

void QStringConverter_ResetState(QStringConverter* self) {
	self->resetState();
}

bool QStringConverter_HasError(const QStringConverter* self) {
	return self->hasError();
}

const char* QStringConverter_Name(const QStringConverter* self) {
	return self->name();
}

const char* QStringConverter_NameForEncoding(int e) {
	return QStringConverter::nameForEncoding(static_cast<QStringConverter::Encoding>(e));
}

libqt_list QStringConverter_AvailableCodecs() {
	auto _ret = QStringConverter::availableCodecs();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

QStringConverterBase__State* QStringConverterBase__State_new() {
	 return new QStringConverterBase::State();
}

QStringConverterBase__State* QStringConverterBase__State_new2(int f) {
	 return new QStringConverterBase::State(static_cast<QFlags<QTextOption::Flag>>(f));
}

void QStringConverterBase__State_Clear(QStringConverterBase__State* self) {
	self->clear();
}

void QStringConverterBase__State_Reset(QStringConverterBase__State* self) {
	self->reset();
}

int QStringConverterBase__State_Flags(const QStringConverterBase__State* self) {
	return self->flags;
}

void QStringConverterBase__State_SetFlags(QStringConverterBase__State* self, int flags) {
	self->flags;
}

int QStringConverterBase__State_InternalState(const QStringConverterBase__State* self) {
	return self->internalState;
}

void QStringConverterBase__State_SetInternalState(QStringConverterBase__State* self, int internalState) {
	self->internalState;
}

ptrdiff_t QStringConverterBase__State_RemainingChars(const QStringConverterBase__State* self) {
	return self->remainingChars;
}

void QStringConverterBase__State_SetRemainingChars(QStringConverterBase__State* self, ptrdiff_t remainingChars) {
	self->remainingChars;
}

ptrdiff_t QStringConverterBase__State_InvalidChars(const QStringConverterBase__State* self) {
	return self->invalidChars;
}

void QStringConverterBase__State_SetInvalidChars(QStringConverterBase__State* self, ptrdiff_t invalidChars) {
	self->invalidChars;
}

void QStringConverterBase__State_Delete(QStringConverterBase__State* self) {
    delete self;
}

File: binding/include/libqstringconverter_base.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qstringconverter.html
pub const qstringconverter = struct {

    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QStringConverter_IsValid(@ptrCast(self));
    }

    pub fn ResetState(self: ?*anyopaque, ) void {
        qtc.QStringConverter_ResetState(@ptrCast(self));
    }

    pub fn HasError(self: ?*anyopaque, ) bool {
        return qtc.QStringConverter_HasError(@ptrCast(self));
    }

    pub fn Name(self: ?*anyopaque, ) []const u8 {
        const _ret = qtc.QStringConverter_Name(@ptrCast(self));
return  std.mem.span(_ret);
    }

    pub fn NameForEncoding(e: i32) []const u8 {
        const _ret = qtc.QStringConverter_NameForEncoding(@intCast(e));
return  std.mem.span(_ret);
    }

    pub fn AvailableCodecs() []const u8 {
        return qtc.QStringConverter_AvailableCodecs();
    }
};

/// https://doc.qt.io/qt-6/qstringconverterbase-state.html
pub const qstringconverterbase__state = struct {

    /// New constructs a new QStringConverterBase::State object.
    pub fn New() QtC.QStringConverterBase__State {
        return qtc.QStringConverterBase__State_new();
    }


    /// New2 constructs a new QStringConverterBase::State object.
    pub fn New2(f: i64) QtC.QStringConverterBase__State {
        return qtc.QStringConverterBase__State_new2(@intCast(f));
    }


    pub fn Clear(self: ?*anyopaque, ) void {
        qtc.QStringConverterBase__State_Clear(@ptrCast(self));
    }

    pub fn Reset(self: ?*anyopaque, ) void {
        qtc.QStringConverterBase__State_Reset(@ptrCast(self));
    }

    pub fn Flags(self: ?*anyopaque, ) i32 {
        return qtc.QStringConverterBase__State_Flags(@ptrCast(self));
    }

    pub fn SetFlags(self: ?*anyopaque, flags: i32) void {
        qtc.QStringConverterBase__State_SetFlags(@ptrCast(self), @intCast(flags));
    }

    pub fn InternalState(self: ?*anyopaque, ) i32 {
        return qtc.QStringConverterBase__State_InternalState(@ptrCast(self));
    }

    pub fn SetInternalState(self: ?*anyopaque, internalState: i32) void {
        qtc.QStringConverterBase__State_SetInternalState(@ptrCast(self), internalState);
    }

    pub fn RemainingChars(self: ?*anyopaque, ) i64 {
        return qtc.QStringConverterBase__State_RemainingChars(@ptrCast(self));
    }

    pub fn SetRemainingChars(self: ?*anyopaque, remainingChars: i64) void {
        qtc.QStringConverterBase__State_SetRemainingChars(@ptrCast(self), remainingChars);
    }

    pub fn InvalidChars(self: ?*anyopaque, ) i64 {
        return qtc.QStringConverterBase__State_InvalidChars(@ptrCast(self));
    }

    pub fn SetInvalidChars(self: ?*anyopaque, invalidChars: i64) void {
        qtc.QStringConverterBase__State_SetInvalidChars(@ptrCast(self), invalidChars);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QStringConverterBase__State_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/stringconverter_base.html#types
pub const enums = struct {
    pub const Flag = enum {
        pub const Default: i32 = 0;
        pub const Stateless: i32 = 1;
        pub const ConvertInvalidToNull: i32 = 2;
        pub const WriteBom: i32 = 4;
        pub const ConvertInitialBom: i32 = 8;
        pub const UsesIcu: i32 = 16;
    };

    pub const Encoding = enum {
        pub const Utf8: i32 = 0;
        pub const Utf16: i32 = 1;
        pub const Utf16LE: i32 = 2;
        pub const Utf16BE: i32 = 3;
        pub const Utf32: i32 = 4;
        pub const Utf32LE: i32 = 5;
        pub const Utf32BE: i32 = 6;
        pub const Latin1: i32 = 7;
        pub const System: i32 = 8;
        pub const LastEncoding: i32 = 8;
    };

};
